### PR TITLE
HTCONDOR-1087 - V9_9-HTCONDOR-1087-pcre2

### PR DESCRIPTION
**Pull request by GregThain**
https://opensciencegrid.atlassian.net/browse/HTCONDOR-1087

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
In regexp_helper(), make the following changes:
* Change the type of 'options' and remove 'options_pcre2'
* Change the type of 'target_idx' and remove casts
* Remove the cast when freeing groups_pcre2

**Comment by GregThain** [comment]
I don't know that there's a clean solution for the `pcre2_substring_list_free` cast -- it takes a `PCRE2_SPTR *`, but `pcre2_substring_list_get` takes a PCRE2_UCHAR ***, and returns in it a list of matches in one allocated block that pcre2_substring_list_free returns: https://www.pcre.org/current/doc/html/pcre2_substring_list_get.html

**Comment by JaimeFrey** [comment]
I haven't looked at the actual header files, but PCRE2_SPTR and PCRE2_UCHAR  aren't unsigned char* and unsigned char? That sounds like a bad API if it requires casting.

**Comment by JaimeFrey** [comment]
Ugh. PCRE2_SPTR is const unsigned char*. So without the explicit cast, the pcre2_substring_list_free() requires an implicit cast from unsigned char** to const unsigned char**, which can violate const protection in the general case.

**Review by JaimeFrey — APPROVED** [review]
LGTM

File: src/condor_starter.V6.1/starter.cpp [hunk 1]
```diff
@@ -3507,10 +3509,11 @@ static void SetEnvironmentForAssignedRes(Env* proc_env, const char * proto, cons
...
 		} else {
 			const char * resid;
-			int cGroups = 0;
-			pcre_fullinfo(re, NULL, PCRE_INFO_CAPTURECOUNT, &cGroups);
-			int ovecsize = 3 * (cGroups + 1); // +1 for the string itself
-			int * ovector = (int *) malloc(ovecsize * sizeof(int));
+			//int cGroups = 0;
+			//pcre_fullinfo(re, NULL, PCRE_INFO_CAPTURECOUNT, &cGroups);
+			//int ovecsize = 3 * (cGroups + 1); // +1 for the string itself
+			//int * ovector = (int *) malloc(ovecsize * sizeof(int));
```
Comment (JaimeFrey) [inline src/condor_starter.V6.1/starter.cpp:3515]:
```suggestion
```

File: src/condor_starter.V6.1/starter.cpp [hunk 2]
```diff
@@ -3519,9 +3522,12 @@ static void SetEnvironmentForAssignedRes(Env* proc_env, const char * proto, cons
 			while ((resid = ids.next())) {
 				if ( ! rhs.empty()) { rhs += env_id_separator; }
 				int cchresid = (int)strlen(resid);
-				int status = pcre_exec(re, NULL, resid, cchresid, 0, 0, ovector, ovecsize);
+				PCRE2_SPTR resid_pcre2str = reinterpret_cast<const unsigned char *>(resid);
+				//int status = pcre_exec(re, NULL, resid, cchresid, 0, 0, ovector, ovecsize);
```
Comment (JaimeFrey) [inline src/condor_starter.V6.1/starter.cpp:3526]:
```suggestion
```

File: src/condor_starter.V6.1/starter.cpp [hunk 3]
```diff
@@ -3539,10 +3545,11 @@ static void SetEnvironmentForAssignedRes(Env* proc_env, const char * proto, cons
 					rhs += resid;
 				}
 			}
-			free(ovector);
+			//free(ovector);
```
Comment (JaimeFrey) [inline src/condor_starter.V6.1/starter.cpp:3548]:
```suggestion
```

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
Remove old commented-out code.
Otherwise, looks good.

**Comment by JaimeFrey** [comment]
Does this merit a mention in the version history?

**Review by JaimeFrey — APPROVED** [review]
I should make sure to hit the "Approve" radio button when I approve a PR.

**Pull request merged**